### ISSUE TITLE
Maybe you can use this for comparision 

### ISSUE THREAD

**Issue by donnerbrenn** [blackle/Megalania#1]
Hi, Blackle,

nice idea, and i don't mean to belittle your efforts. But I wrote a little python3 script, that seems to perform out this by a few bytes on international shipping (and everything else I tried), just by bruteforcing the nice value... Maybe you can use it for comparision purposes with Megalania.
[nicer.zip](https://github.com/blackle/Megalania/files/4431128/nicer.zip)

Use it with 
./nicer.py in.file -o out.file

Shipping packed with...
Latest Megalania: 3795 Bytes
Nicer: 3782 Bytes

(I also applied some more compiler voodoo to the Makefile of shipping, to reduce the filesize further...)

**Comment by blackle** [comment]
thanks! yeah it seems like simulated annealing can't really get close to the true optimal on its own. I was going to explore other search methods to find the best. alternatively, I might make it so you can pass in an existing compressed file and it will just improve it.

**Comment by blackle** [comment]
also, could you share this "voodoo" you did to make international shipping smaller? I'd love to know

**Comment by donnerbrenn** [comment]
Sure... Actually, I hoped, you would ask. :-D

I quickly created a git-repo... But I didn't clean up... So there might be some unnecessary files included from my experiments today.
https://github.com/donnerbrenn/international-shipping

Just look at the CFLAGS, I've added. 

Also there is a linker script now, that squeezesout some more bytes. Actually, it blows up the unpacked executable to 17MB, but this is gone after the strip command... Pretty sure, I'm doing something wrong there... But for now it works.

With gcc 9.3 it's at 3664 Bytes now (Without the timed rendering #define enabled.). If you are experiencing troubles with your compiler (you mention gcc6 in your Makefile... LTO doesn't seem to be supported correctly there or at least different), just  comment out line 48 of the Makefile (CFLAGS+=-flto).

> I might make it so you can pass in an existing compressed file and it will just improve it.

I have no idea, if this can work but... Maybe it's worth a try.

**Comment by donnerbrenn** [comment]
Oh... and... I changed "march" to nocona. Makes no difference in the compiled size, but seems to compress better than core2.

**Comment by blackle** [comment]
This is awesome!!! thank you so much!

**Comment by blackle** [comment]
Also, because of the way xzutils compresses files there will always be at least 5 bytes that can be saved. This is because the "uncompressed size" field is not used by xzutils, and it instead uses a special "end of stream" symbol which takes about 5 bytes. Megalania uses the uncompressed size field, so at the very least it could be a post-processor lol

**Comment by donnerbrenn** [comment]
Haha! Every byte counts.

You're welcome. I've learned so much from your OGL-Templates and so on. So I finally could give something back. Maybe you can use some of this stuff in your next scene-production. ;)

Btw.: Looking forward to this talk by Unlord. Sounds, like he could release a usable version of XLINK. I'm waiting, since ShizMob and PoroCyon talked about this on last years Revision. Crinkler like compression on Linux would be just awesome.
https://2020.revision-party.net/events/seminars#unlord

**Comment by blackle** [comment]
oh my gosh it makes me very happy that the templates helped! I'll definitely use your stuff in my next prod... which could be for revision if I actually do something for it :fearful: 

I'm super hyped for unlord's talk. I'd be glad if xlink obsoletes lzma, but I have a feeling that lzma might win out for very small demo sizes like 1k or 512b, just by virtue of being available on every linux system

**Comment by donnerbrenn** [comment]
Would love to see a new 4K thingy from you. ;)

I never released anything. ^^

**Comment by donnerbrenn** [comment]
Hmmmm.... I tried to apply this to nicer

> Also, because of the way xzutils compresses files there will always be at least 5 bytes that can be saved. This is because the "uncompressed size" field is not used by xzutils, and it instead uses a special "end of stream" symbol which takes about 5 bytes. Megalania uses the uncompressed size field, so at the very least it could be a post-processor lol

So i took this specification and nicer and wrote a little function, that's supposed to "fix" this.
https://github.com/jljusten/LZMA-SDK/blob/master/DOC/lzma-specification.txt

Basicly I write the size of the resulting binary to the LZMA-header and truncate the last 6 Bytes of the archive. For my understanding the end marker is located there. XZ now complains, the compressed data is corrupted, but decompresses anyway. Actually it complains already after changing the header... Without truncating. I don't get why.... But I will figure this out sooner or later.

Anyway... It works. Down to 3658 Bytes now. I've updated my repo.
